### ISSUE TITLE
raft topology: `system.peers` contains partial rows (only key and host ID) for non-normal nodes, which confuses (some) drivers

### ISSUE THREAD

**Issue by kbr-scylla** [scylladb/scylladb#19507]
In 6.0, in raft-topology mode, we (ab)use `system.peers` to persist IP mappings of nodes that have not yet finished bootstrapping -- nodes not in NORMAL state.

But `system.peers` acts as an interface for the drivers -- they fetch it from control connection and use it (together with `system.local`) to get a view of the cluster's topology.

A row in `system.peers` is considered invalid by drivers if it misses some columns. For example, here's the logic from Python driver:
https://github.com/scylladb/python-driver/blob/c9b24b74841f5aecc5dac6f94454950a0ebf76ac/cassandra/cluster.py#L4018

Python driver handles such rows gracefully -- prints a warning, but otherwise constructs topology metadata based on rows that it considers valid. (`_refresh_node_list_and_token_map` function)

However other drivers (at least one -- Rust driver) completely abandon processing `system.peers` if they encounter invalid rows, as observed in https://github.com/scylladb/scylladb/issues/19107#issuecomment-2186639040

There are two potential solutions to consider:
1. change in Scylla: do not use `system.peers` to store mappings of non-NORMAL nodes. We'd have to introduce a separate table for that.
2. or change in drivers: make sure all drivers handle the rows gracefully, ignoring them, but otherwise handling the rest of `system.peers` correctly.

(2) might need to be done anyway as part of https://github.com/scylladb/scylladb/issues/6527 -- which will introduce NORMAL nodes with 0 tokens. Now the question is, how a node with 0 tokens appears in `system.peers` -- is the column `null`, or is it a non-`null` empty set? (IIRC this depends on whether the collection is atomic or not). cc @patjed41 

Another question: how many drivers actually have a problem here. We know that Rust driver does, but this is the only driver (for now) that we did not inherit from Cassandra. Perhaps all Cassandra drivers already handle this gracefully -- and by extension, our forks of those drivers do too.

If it turned out that some Cassandra driver does not handle this gracefully, that would push us in the direction of (1); OTOH we need (2) for #6527 as mentioned before (in which case we'd have to stop supporting this particular Cassandra driver, forcing everyone to migrate to Scylla driver, if they use 0 token nodes)

---

Another question which appeared in #19107 is when do we deliver notifications about nodes joining. @Lorak-mmk mentioned that we must only send notification about node that is fully ready (i.e. NORMAL)

**Comment by gleb-cloudius** [comment]
> change in Scylla: do not use system.peers to store mappings of non-NORMAL nodes. We'd have to introduce a separate table for that.

May be scylla_local can be used for that.

**Comment by kbr-scylla** [comment]
> Another question which appeared in https://github.com/scylladb/scylladb/issues/19107 is when do we deliver notifications about nodes joining. @Lorak-mmk mentioned that we must only send notification about node that is fully ready (i.e. NORMAL)

```cpp
void cql_server::event_notifier::send_join_cluster(const gms::inet_address& endpoint)
{
    for (auto&& conn : _topology_change_listeners) {
        using namespace cql_transport;
        if (!conn->_pending_requests_gate.is_closed()) {
            conn->write_response(conn->make_topology_change_event(event::topology_change::new_node(endpoint, conn->_server_addr.port())));
        };
    }
}
```

called from

```cpp
void cql_server::event_notifier::on_join_cluster(const gms::inet_address& endpoint)
{
    if (!_server._gossiper.is_cql_ready(endpoint)) {
        _endpoints_pending_joined_notification.insert(endpoint);
        return;
    }

    send_join_cluster(endpoint);
}
```

called from
```cpp
future<> endpoint_lifecycle_notifier::notify_joined(gms::inet_address endpoint) {
    return seastar::async([this, endpoint] {
        _subscribers.thread_for_each([endpoint] (endpoint_lifecycle_subscriber* subscriber) {
            try {
                subscriber->on_join_cluster(endpoint);
            } catch (...) {
                slogger.warn("Join cluster notification failed {}: {}", endpoint, std::current_exception());
            }
        });
    });
}
```

called from

```cpp
future<> storage_service::notify_joined(inet_address endpoint) {
    co_await utils::get_local_injector().inject(
        "storage_service_notify_joined_sleep", std::chrono::milliseconds{500});

    co_await container().invoke_on_all([endpoint] (auto&& ss) {
        return ss._lifecycle_notifier.notify_joined(endpoint);
    });
    slogger.debug("Notify node {} has joined the cluster", endpoint);
}
```

called from
```cpp
future<> storage_service::notify_nodes_after_sync(nodes_to_notify_after_sync&& nodes_to_notify) {
    for (auto [ip, host_id] : nodes_to_notify.left) {
        co_await notify_left(ip, host_id);
    }
    for (auto ip : nodes_to_notify.joined) {
        co_await notify_joined(ip);
    }
}
```


called in two places:
1. `storage_service::raft_ip_address_updater`
```cpp
                storage_service::nodes_to_notify_after_sync nodes_to_notify;
                auto lock = co_await _ss.get_token_metadata_lock();
                co_await _ss.mutate_token_metadata([this, hid, &nodes_to_notify](mutable_token_metadata_ptr t) -> future<> {
                    nodes_to_notify = co_await _ss.sync_raft_topology_nodes(std::move(t), hid, {});
                }, storage_service::acquire_merge_lock::no);
                co_await _ss.notify_nodes_after_sync(std::move(nodes_to_notify));
```
2. `storage_service::topology_state_load`
```cpp
        auto nodes_to_notify = co_await sync_raft_topology_nodes(tmptr, std::nullopt, std::move(prev_normal));

        if (_db.local().get_config().enable_tablets()) {
            tmptr->set_tablets(co_await replica::read_tablet_metadata(_qp));
            tmptr->tablets().set_balancing_enabled(_topology_state_machine._topology.tablet_balancing_enabled);
        }

        co_await replicate_to_all_cores(std::move(tmptr));
        co_await notify_nodes_after_sync(std::move(nodes_to_notify));
```

and `sync_raft_topology_nodes` adds a node to `nodes_to_notify.joined` **only when it becomes NORMAL** (in `process_normal_node`)

so regarding driver notifications about joining nodes -- we're good.

**Comment by mykaul** [comment]
> > change in Scylla: do not use system.peers to store mappings of non-NORMAL nodes. We'd have to introduce a separate table for that.
> 
> May be scylla_local can be used for that.

Let's verify beforehand that the drivers don't use that too in a similar manner.  From https://github.com/scylladb/gocql/blob/74675d1c5ba516724eb09732cac9a7abc2fb9936/control.go#L270 : 
```golang
func (c *controlConn) setupConn(conn *Conn) error {
	// we need up-to-date host info for the filterHost call below
	iter := conn.querySystemLocal(context.TODO())
	defaultPort := 9042
	if tcpAddr, ok := conn.conn.RemoteAddr().(*net.TCPAddr); ok {
		defaultPort = tcpAddr.Port
	}
	host, err := c.session.hostInfoFromIter(iter, conn.host.connectAddress, defaultPort)
	if err != nil {
		return err
	}
```

**Comment by kbr-scylla** [comment]
Regarding drivers, we know that Python driver and gocql handle invalid peers entries gracefully.

Java driver 3.x:
```java
  private boolean isValidPeer(Row peerRow, boolean logIfInvalid) {
    boolean isValid =
        peerRow.getColumnDefinitions().contains("host_id") && !peerRow.isNull("host_id");

    if (isPeersV2) {
      isValid &=
          peerRow.getColumnDefinitions().contains("native_address")
              && peerRow.getColumnDefinitions().contains("native_port")
              && !peerRow.isNull("native_address")
              && !peerRow.isNull("native_port");
    } else {
      isValid &=
          (peerRow.getColumnDefinitions().contains("rpc_address") && !peerRow.isNull("rpc_address"))
              || (peerRow.getColumnDefinitions().contains("native_transport_address")
                  && peerRow.getColumnDefinitions().contains("native_transport_port")
                  && !peerRow.isNull("native_transport_address")
                  && !peerRow.isNull("native_transport_port"));
    }

    if (EXTENDED_PEER_CHECK) {
      isValid &=
          peerRow.getColumnDefinitions().contains("data_center")
              && !peerRow.isNull("data_center")
              && peerRow.getColumnDefinitions().contains("rack")
              && !peerRow.isNull("rack")
              && peerRow.getColumnDefinitions().contains("tokens")
              && !peerRow.isNull("tokens");
    }
    if (!isValid && logIfInvalid)
      logger.warn(
          "Found invalid row in system.peers: {}. "
              + "This is likely a gossip or snitch issue, this host will be ignored.",
          formatInvalidPeer(peerRow));
    return isValid;
  }
```

in `refreshNodeInfo`:
```java
        // Ignore hosts with a null rpc_address, as this is most likely a phantom row in
        // system.peers (JAVA-428).
        // Don't test this for the control host since we're already connected to it anyway, and we
        // read the info from system.local
        // which didn't have an rpc_address column (JAVA-546) until CASSANDRA-9436
      } else if (!c.endPoint.equals(host.getEndPoint()) && !isValidPeer(row, true)) {
        return false;
```

(rpc_address reference spotted! cc @fee-mendes)


Rust driver fails.


@Lorak-mmk could you check other drivers?

**Comment by gleb-cloudius** [comment]
> > > change in Scylla: do not use system.peers to store mappings of non-NORMAL nodes. We'd have to introduce a separate table for that.
> > 
> > 
> > May be scylla_local can be used for that.
> 
> Let's verify beforehand that the drivers don't use that too in a similar manner. 

This is generic key/value store. As long as you use unique keys you should not worry about stepping on  someone feet.

**Comment by Lorak-mmk** [comment]
As far as I can tell Python Driver, Java Driver 4.x and gocql skip invalid peers @kbr-scylla , so Rust Drivers behavior is in the wrong here. I opened an issue about this: https://github.com/scylladb/scylla-rust-driver/issues/1023

**Comment by roydahan** [comment]
The issue with rust driver should be fixed and available in 0.14.0 (and then with a new cql-stress to be released with that).
Closing this one and leaving the EPIC to make sure we have tests for it in all other drivers.

**Comment by Lorak-mmk** [comment]
> The issue with rust driver should be fixed and available in 0.14.0 (and then with a new cql-stress to be released with that). Closing this one and leaving the EPIC to make sure we have tests for it in all other drivers.

I think this is already released in 0.13.2. @wprzytula

**Comment by roydahan** [comment]
Correct, I forgot we released 0.13.2
